File: backend/app/api/v1/endpoints/memory_cases.py
"""Memory case browsing endpoints."""

from uuid import UUID

from fastapi import APIRouter, Depends, Query, Request
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.responses import success_response
from app.db.session import get_db_session
from app.repositories.memory_case import MemoryCaseRepository
from app.schemas.common import SuccessResponse
from app.schemas.memory_case import MemoryCaseCollection, MemoryCaseRead

router = APIRouter(prefix="/memory-cases", tags=["memory-cases"])


@router.get("", response_model=SuccessResponse[MemoryCaseCollection])
async def list_memory_cases(
    request: Request,
    region_id: UUID | None = Query(default=None),
    station_id: UUID | None = Query(default=None),
    severity: str | None = Query(default=None),
    q: str | None = Query(default=None, description="Search objective, summary, keywords and payloads."),
    limit: int = Query(default=24, ge=1, le=200),
    session: AsyncSession = Depends(get_db_session),
):
    """List recent memory cases for the dedicated UI page."""
    memory_cases = await MemoryCaseRepository(session).list_recent(
        limit=limit,
        region_id=region_id,
        station_id=station_id,
        severity=severity,
        query=q,
    )
    return success_response(
        request=request,
        message="Memory cases retrieved successfully.",
        data=MemoryCaseCollection(
            items=[MemoryCaseRead.model_validate(memory_case) for memory_case in memory_cases],
            count=len(memory_cases),
        ),
    )

File: frontend/src/lib/api/memoryCases.ts
import { apiGet } from "./http";

export interface MemoryCaseRead {
  id: string;
  created_at: string;
  updated_at: string;
  region_id: string;
  station_id: string | null;
  risk_assessment_id: string | null;
  incident_id: string | null;
  action_plan_id: string | null;
  action_execution_id: string | null;
  decision_log_id: string | null;
  objective: string | null;
  severity: string | null;
  outcome_class: string;
  outcome_status_legacy: string | null;
  summary: string;
  context_payload: Record<string, unknown> | null;
  action_payload: Record<string, unknown> | null;
  outcome_payload: Record<string, unknown> | null;
  keywords: string[] | null;
  occurred_at: string;
}

export interface MemoryCaseCollection {
  items: MemoryCaseRead[];
  count: number;
}

export function getMemoryCases(
  query?: {
    region_id?: string;
    station_id?: string;
    severity?: string;
    q?: string;
    limit?: number;
  },
  signal?: AbortSignal,
): Promise<MemoryCaseCollection> {
  return apiGet<MemoryCaseCollection>("/memory-cases", { query, signal });
}